File: testing/contracts/cw-ica-owner/src/msg.rs
use cosmwasm_schema::{cw_serde, QueryResponses};
use cw_ica_controller::helpers::ica_callback_execute;
use cw_ica_controller::types::msg::options::ChannelOpenInitOptions;

#[cw_serde]
pub struct InstantiateMsg {
    pub admin: Option<String>,
    pub ica_controller_code_id: u64,
}

#[ica_callback_execute]
#[cw_serde]
pub enum ExecuteMsg {
    CreateIcaContract {
        salt: Option<String>,
        channel_open_init_options: ChannelOpenInitOptions,
    },
    /// SendPredefinedAction sends a predefined action from the ICA controller to the ICA host.
    /// This demonstration is useful for contracts that have predefined actions such as DAOs.
    ///
    /// In this example, the predefined action is a `MsgSend` message which sends 100 "stake" tokens.
    SendPredefinedAction {
        /// The ICA ID.
        ica_id: u64,
        /// The recipient's address, on the counterparty chain, to send the tokens to from ICA host.
        to_address: String,
    },
}

#[cw_serde]
#[derive(QueryResponses)]
pub enum QueryMsg {
    /// GetContractState returns the contact's state.
    #[returns(crate::state::ContractState)]
    GetContractState {},
    /// GetIcaState returns the ICA state for the given ICA ID.
    #[returns(crate::state::IcaContractState)]
    GetIcaContractState { ica_id: u64 },
    /// GetIcaCount returns the number of ICAs.
    #[returns(u64)]
    GetIcaCount {},
}

File: e2e/interchaintest/types/owner/msg.go
package owner

import (
	"encoding/json"

	"github.com/srdtrk/cw-ica-controller/interchaintest/v2/types/icacontroller"
)

type InstantiateMsg struct {
	// The admin address. If not specified, the sender is the admin.
	Admin               *string `json:"admin,omitempty"`
	IcaControllerCodeId uint64  `json:"ica_controller_code_id"`
}

type ExecuteMsg struct {
	CreateIcaContract    *ExecuteMsg_CreateIcaContract    `json:"create_ica_contract,omitempty"`
	SendPredefinedAction *ExecuteMsg_SendPredefinedAction `json:"send_predefined_action,omitempty"`
}

type ExecuteMsg_CreateIcaContract struct {
	Salt                   *string                              `json:"salt,omitempty"`
	ChannelOpenInitOptions icacontroller.ChannelOpenInitOptions `json:"channel_open_init_options,omitempty"`
}

type ExecuteMsg_SendPredefinedAction struct {
	IcaId     uint64 `json:"ica_id"`
	ToAddress string `json:"to_address"`
}

type QueryMsg struct {
	// GetContractState returns the contract state
	GetContractState    *struct{}                     `json:"get_contract_state,omitempty"`
	GetIcaContractState *QueryMsg_GetIcaContractState `json:"get_ica_contract_state,omitempty"`
	GetIcaCount         *struct{}                     `json:"get_ica_count,omitempty"`
}

type QueryMsg_GetIcaContractState struct {
	IcaId uint64 `json:"ica_id"`
}

// ToString returns a string representation of the message
func (m *InstantiateMsg) ToString() string {
	return toString(m)
}

// ToString returns a string representation of the message
func (m *ExecuteMsg) ToString() string {
	return toString(m)
}

// ToString returns a string representation of the message
func (m *QueryMsg) ToString() string {
	return toString(m)
}

func toString(v any) string {
	jsonBz, err := json.Marshal(v)
	if err != nil {
		panic(err)
	}

	return string(jsonBz)
}
